File: lang/java/src/fun/vac/data_structures/ArrayList.java
package fun.vac.data_structures;

import fun.vac.data_structures.interfaces.IList;

public class ArrayList<E> implements IList<E> {

    private static final int DEFAULT_CAPACITY = 64;

    private E[] data = null;
    private int logicalSize = 0;
    private int physicalSize = DEFAULT_CAPACITY;

    public ArrayList() {
        this(0);
    }

    @SuppressWarnings("unchecked")
    public ArrayList(int physicalSize) {
        if (physicalSize > 1) {
            this.physicalSize = physicalSize;
        }
        data = (E[]) new Object[this.physicalSize];
    }

    @Override
    public int size() {
        return logicalSize;
    }

    @Override
    public boolean isEmpty() {
        return logicalSize == 0;
    }

    @Override
    public E get(int index) {
        if (index < 0 || index >= logicalSize) {
            throw new RuntimeException("[PANIC - IndexOutOfBounds]");
        }

        return data[index];
    }

    @Override
    public void set(int index, E element) {
        if (index < 0 || index >= logicalSize) {
            throw new RuntimeException("[PANIC - IndexOutOfBounds]");
        }

        data[index] = element;
    }

    @Override
    @SuppressWarnings("unchecked")
    public void insert(int index, E element) {
        if (index < 0 || index > logicalSize) {
            throw new RuntimeException("[PANIC - IndexOutOfBounds]");
        }

        if (logicalSize == physicalSize) {
            int newCapacity = DEFAULT_CAPACITY;

            if (physicalSize >= DEFAULT_CAPACITY) {
                newCapacity = physicalSize + (physicalSize >>> 1);
            }

            E[] temp = (E[]) new Object[newCapacity];

            for (int i = 0, length = logicalSize; i < length; i++) {
                temp[i] = data[i];
            }

            data = temp;
            physicalSize = newCapacity;
        }

        for (int i = logicalSize; i > index; i--) {
            data[i] = data[i - 1];
        }

        data[index] = element;

        logicalSize++;
    }

    @Override
    public void remove(int index) {
        if (index < 0 || index >= logicalSize) {
            throw new RuntimeException("[PANIC - IndexOutOfBounds]");
        }

        for (int i = index + 1; i < logicalSize; i++) {
            data[i - 1] = data[i];
        }

        logicalSize--;

        data[logicalSize] = null;
    }

    @Override
    public E front() {
        return get(0);
    }

    @Override
    public E back() {
        return get(logicalSize - 1);
    }

    @Override
    public void prepend(E element) {
        insert(0, element);
    }

    @Override
    public void append(E element) {
        insert(logicalSize, element);
    }

    @Override
    public void poll() {
        remove(0);
    }

    @Override
    public void eject() {
        remove(logicalSize - 1);
    }

    public int capacity() {
        return physicalSize;
    }

    @SuppressWarnings("unchecked")
    public void shrink() {
        E[] temp = (E[]) new Object[logicalSize];

        for (int i = 0, length = logicalSize; i < length; i++) {
            temp[i] = data[i];
        }

        data = temp;
        physicalSize = logicalSize;
    }
}

File: lang/php/src/fun/vac/data_structures/ArrayList.php
<?php
    namespace fun\vac\data_structures\ArrayList;

    require_once "fun/vac/data_structures/interfaces/IList.php";

    use fun\vac\data_structures\interfaces\IList;

    class ArrayList implements IList\IList {
        private const DEFAULT_CAPACITY = 64;

        private $data = NULL;
        private $logicalSize = 0;
        private $physicalSize = self::DEFAULT_CAPACITY;

        public function __construct(int $physicalSize = 0) {
            if ($physicalSize > 1) {
                $this->physicalSize = $physicalSize;
            }
            $this->data = array_pad(array(), $this->physicalSize, 0);
        }

        public function size(): int {
            return $this->logicalSize;
        }

        public function isEmpty(): bool {
            return $this->logicalSize == 0;
        }

        public function get(int $index): int {
            if ($index < 0 || $index >= $this->logicalSize) {
                throw new RuntimeException("[PANIC - IndexOutOfBounds]");
            }

            return $this->data[$index];
        }

        public function set(int $index, int $element): void {
            if ($index < 0 || $index >= $this->logicalSize) {
                throw new RuntimeException("[PANIC - IndexOutOfBounds]");
            }

            $this->data[$index] = $element;
        }

        public function insert(int $index, int $element): void {
            if ($index < 0 || $index > $this->logicalSize) {
                throw new RuntimeException("[PANIC - IndexOutOfBounds]");
            }

            if ($this->logicalSize == $this->physicalSize) {
                $newCapacity = self::DEFAULT_CAPACITY;

                if ($this->physicalSize >= self::DEFAULT_CAPACITY) {
                    $newCapacity = $this->physicalSize + ($this->physicalSize >> 1);
                }

                $temp = array_pad(array(), $newCapacity, 0);

                for ($i = 0, $length = $this->logicalSize; $i < $length; $i++) {
                    $temp[$i] = $this->data[$i];
                }

                $this->data = $temp;
                $this->physicalSize = $newCapacity;
            }

            for ($i = $this->logicalSize; $i > $index; $i--) {
                $this->data[$i] = $this->data[$i - 1];
            }

            $this->data[$index] = $element;

            $this->logicalSize++;
        }

        public function remove(int $index): void {
            if ($index < 0 || $index >= $this->logicalSize) {
                throw new RuntimeException("[PANIC - IndexOutOfBounds]");
            }

            for ($i = $index + 1; $i < $this->logicalSize; $i++) {
                $this->data[$i - 1] = $this->data[$i];
            }

            $this->logicalSize--;

            $this->data[$this->logicalSize] = NULL;
        }

        public function front(): int {
            return $this->get(0);
        }

        public function back(): int {
            return $this->get($this->logicalSize - 1);
        }

        public function prepend(int $element): void {
            $this->insert(0, $element);
        }

        public function append(int $element): void {
            $this->insert($this->logicalSize, $element);
        }

        public function poll(): void {
            $this->remove(0);
        }

        public function eject(): void {
            $this->remove($this->logicalSize - 1);
        }

        public function capacity(): int {
            return $this->physicalSize;
        }

        public function shrink(): void {
            $temp = array_pad(array(), $this->logicalSize, 0);

            for ($i = 0, $length = $this->logicalSize; $i < $length; $i++) {
                $temp[$i] = $this->data[$i];
            }

            $this->data = $temp;
            $this->physicalSize = $this->logicalSize;
        }
    }
?>
